docs/quantum_gate.rst

```diff
@@ -1,4 +1,366 @@
+and is a structural attribute of the kernel object.
+The gates in this vector are assumed to be executed from the first to the last in the vector.
+
+During a scheduling pass, the ``cycle`` attribute of each gate gets defined.
+See its definition in :ref:`quantum_gate_attributes_in_the_internal_representation`.
+The gates in the vector then are ordered in non-decreasing cycle order.
+
+The schedulers also produce a ``bundled`` version of each circuit.
+The circuit is then represented by a list of bundles
+in which each bundle represents the gates that are to be started in a particular cycle.
+Each bundle is structured as a list of sections and each section as a list of gates (actually gate pointers).
+The gates in each section share the same operation but have different operands, obviously.
+The latter prepares for code generation for a SIMD instruction set
+in which a single instruction with one operation can have multiple operands.
+Each bundle has two additional attributes:
+
+- ``start_cycle`` representing the cycle in which all gates of the bundle start
+
+- ``duration_in_cycles`` representing the maximum duration in cycles of the gates in the bundle
+
+This internal bundles representation is used during QISA generation instead of the original circuit.
+
+
+.. _input_external_representation:
+
+Input external representation
+-----------------------------
+
+OpenQL supports as input external representation currently only the OpenQL program, written in C++ and/or Python.
+This is an API-level interface based on platform, program, kernel and gate objects and their methods.
+Calls to these methods transfer the external representation into the internal representation
+(also called intermediate representation or IR) as sketched above:
+a program (object) consisting of a vector of kernels,
+each containing a single circuit,
+each circuit being a vector of gates.
+
+Gates are created using an API of the general form:
+
+.. code::
+
+    k.gate(name, qubit operand vector, classical operand vector, duration, angle)
+
+in which particular operands can be empty or 0 depending on the particular kind of gate that is created.
+Gate creation upon a call to this API goes through the following steps to create the internal representation:
+
+#. the qubit and/or classical register operand indices are checked for validity,
+   i.e. to be in the range of 0 to the number specified in the program creation API minus 1
+
+#. if the configuration file contains a definition for a specialized composite gate matching it, it is taken;
+   the qubit parameter substitution in the gates of the decomposition specification is done;
+   each resulting gate must be available as (specialized or parameterized, and non-composite) custom gate,
+   or as a default gate; the decomposition is applied and all resulting gates are created and added to the circuit
+
+#. otherwise, if a parameterized composite gate is available, take it;
```

Comment:
Something went wrong with formatting here. the # is interpreted as a header. Check the rendered docs to see how it looks.